CASSANDRA-12151

vinaykumarchella opened the pull request:
CASSANDRA-12151: AuditLogging in trunk

Commit:
CASSANDRA-12151: AuditLogging in trunk, Added categories for AuditLogEntry types and filtering based on categories.

jaydeepkumar1984 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -0,0 +1,379 @@
+        logger.info("Logger instance loaded with class : " + loggerClassName);
+        try
+        {
+            return (IAuditLogger) Class.forName(loggerClassName).newInstance();
```
Can this code be moved inside util class `FBUtilities.java`? 
Similar to 
`public static IRoleManager newRoleManager(String className)`
`    public static IPartitioner newPartitioner(String partitionerClassName) throws ConfigurationException
`
...

vinaykumarchella replied:
Done

jaydeepkumar1984 commented on src/java/org/apache/cassandra/audit/AuditLogFilter.java:
```diff
@@ -0,0 +1,147 @@
+
+        includedCategories.set(includedCategoriesList);
+        excludedCategories.set(excludedCategoriesList);
+
```
minor comment, can you please remove this extra empty line here?

vinaykumarchella replied:
Done

jaydeepkumar1984 commented on src/java/org/apache/cassandra/transport/messages/BatchMessage.java:
```diff
@@ -237,6 +239,11 @@ public BatchMessage(BatchStatement.Type type, List<Object> queryOrIdList, List<L
             {
                 FullQueryLogger.instance.logBatch(batchType.name(), queryStrings, values, options, fqlTime);
             }
+            if(auditLogEnabled)
```
Here when someone enables audit log then every single query will be audited. In my opinion this should be made flexible by allowing percentage of queries to be audited as some of high throughput system may not be able to tolerate penalty of auditing for every single query.

vinaykumarchella replied:
Ideally, AuditLog would not be percentage based but it would be every statement to meet the stated AuditLog requirements. For data sampling, we might not want to overload AuditLog, since it deviates the actual purpose of AuditLog. However, we can create separate JIRA to start this discussion if you would like.

jaydeepkumar1984 commented on src/java/org/apache/cassandra/transport/messages/BatchMessage.java:
```diff
@@ -245,6 +252,12 @@ public BatchMessage(BatchStatement.Type type, List<Object> queryOrIdList, List<L
         }
         catch (Exception e)
         {
+            if(auditLogEnabled)
```
Can we have some simple API call form here and then in that static call we do all the business check? I am trying to simplify Messaging/QueryProcessing classes to not add much of business logic.

For example, here we can do something like
```

Audit.java
public static shouldAudit(param1, param2,...)
{
    //add feature to not audit all the queries

    if(ThreadLocalRandom.current().nextDouble() > auditFraction)
    //do not audit this query
        return;

    if (auditLogEnabled) 
    {
        ...
    }
}
```

And then from higher classes like `Message.java`, `ExecuteMessage.java.java` we just simply call
`Audit.shouldAudit(.....)`

jolynch replied:
Alternatively we could delegate this entirely to auditLogManager.log and just have any percentage based choices made in the `IAuditLogger`?

vinaykumarchella replied:
Implementers of `IAuditLogger` would have the choice to go with percentage based if they like

Commit:
Implementing code review comments - Moved AuditLogger class initialization to FBUtilities.

Commit:
Code review comment fix: Optimize and simplify ```AuditLogFilter::isFiltered.```

jolynch commented on conf/cassandra.yaml:
```diff
@@ -1168,3 +1168,20 @@ back_pressure_strategy:
+# logger: Class name of the logger/ custom logger
+# includedKeyspaces: comma separated list of keyspaces to be included for audit log, default - includes all keysapces
+# excludedKeyspaces: comma separated list of keyspaces to be excluded for audit log, default - excludes no keysapce
+# includedCategories: comma separated list of Audit Log Categories to be included for audit log, default - includes all categories
```
These should probably be snake_case instead of camel_case.

Once the ticket/code quiets down we can write some docs in the docs tree :-)

Commit:
Fixing code review comment: Simplifying extracting AuditLogType by introducing IAuditLogContext interface, Instead of creating an explicit map of statements, added the type of statement in the actual class itself. That way when new statements are introduced, we don't need to keep updating this map. Removed CL.

Commit:
Fixing code review comment: Added user filtering, updated filtering to snake_case

Commit:
Reusing BinLog for AuditLogging.

Commit:
Added documentation in doc folder

jolynch commented on NEWS.txt:
```diff
@@ -68,6 +68,8 @@ New features
      See nodetool and fqltool help text for more information.
    - SSTableDump now supports the -l option to output each partition as it's own json object
      See CASSANDRA-13848 for more detail
+   - CASSANDRA-12151 Audit logging for database activity. Logs every incoming CQL command request, Authentication
```
nit: "Support audit logging for database activity. If enabled, logs every incoming ..."

vinaykumarchella replied:
Done

jolynch commented on conf/cassandra.yaml:
```diff
@@ -1168,3 +1168,26 @@ back_pressure_strategy:
+
+
+# Audit logging - Logs every incoming CQL command request, Authentication (successful as well as unsuccessful login)
+# to C* node. Currently there are two implementations provided, "FileAuditLogger" which logs events to audit/audit.log file
```
nit: I'd leave out the quotes, just FileAuditLogger and BinAuditLogger

jolynch commented on conf/cassandra.yaml:
```diff
@@ -1168,3 +1168,26 @@ back_pressure_strategy:
+# excluded_users: comma separated list of users to be excluded for audit log, default - excludes no user
+audit_logging_options:
+    enabled: false
+    logger: BinAuditLogger
```
Should BinAuditLogger be the default?

vinaykumarchella replied:
Yes, Considering the performance numbers, I am leaving it as default, but users would have choice to choose what they need

jolynch commented on doc/source/operating/audit_logging.rst:
```diff
@@ -0,0 +1,71 @@
+
+How to configure
+^^^^^^^^^^^^^^^^^
+enabled: This option enables/ disables audit log
```
nit: ``enabled``

jolynch commented on doc/source/operating/audit_logging.rst:
```diff
@@ -0,0 +1,71 @@
+
+What does it log
+^^^^^^^^^^^^^^^^^
+It logs following attributes
```
Maybe something about the final format? E.g. "Each audit log implementation has access to the following attributes, and for the default text based logger these fields are concatenated with |s to yield the final message", or something like that?

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogEntryType.java:
```diff
@@ -0,0 +1,131 @@
+     * CQL Audit Log Entry Types
+     */
+
+    SELECT("QUERY"),
```
I'm not sure how this is going to go when adding new statements. Since we're already adding an interface to the statements themselves perhaps we could just define their AuditLogEntryType there (e.g. this enum becomes just "QUERY", "DML", "DDL", etc?

Maybe this won't work because of the AUTH ones, but it would still be nice to know that new statements change this enum?

vinaykumarchella replied:
Correct, when the new statements get added they would need to implement IAuditLogContext so as part of that they can change this Enum.

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogFilter.java:
```diff
@@ -0,0 +1,156 @@
+
+    public boolean isFiltered(AuditLogEntry auditLogEntry)
+    {
+        return isFiltered(auditLogEntry.getKeyspace(),includedKeyspaces,excludedKeyspaces)
```
nit: add spaces

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -0,0 +1,327 @@
+
+    public void log(CQLStatement statement, String query, QueryOptions options, QueryState state, long queryStartNanoTime)
+    {
+        if(isAuditingEnabled())
```
I like this because it means you can run both the audit logger and the fq logger at the same time, but it might be a good idea to leave a comment here explaining why we have both.

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogUtil.java:
```diff
@@ -0,0 +1,28 @@
+
+public class AuditLogUtil
+{
+    public static final InetSocketAddress DEFAULT_SOURCE = new InetSocketAddress(0);
```
I don't think you want this. InetSocketAddress(0) will be a random port afaik. I think it better to have some kind of "UNKOWN" string or something and then instead of an InetSocketAddress have an Optional<InetSocketAddress>.

vinaykumarchella replied:
I went with Dinesh's suggestion and using "0.0.0.0/0"

jolynch reviewed (commented):
Looks good to me, nice :-)

Commit:
Implemented review suggestions from Dinesh and Joseph

jaydeepkumar1984 reviewed (approved):
Looks good to me

vinaykumarchella committed:
Removed FileAuditLogger references from yaml, added documentation, consider null inputs in AuditLogFilter::isFiltered

laxmikant99 commented on doc/source/operating/audit_logging.rst:
```diff
@@ -0,0 +1,104 @@
+ - ``host``: Host IP, where the command is being executed
+ - ``source ip address``: Source IP address from where the request initiated
+ - ``source port``: Source port number from where the request initiated
+ - ``timestamp``: unix time stamp
```
what is the benefit of timestamp if each log record will have date (%date) ?

laxmikant99 replied:
batch UUID is missing here.

vinaykumarchella replied:
`timestamp` is the only field with date and time details. Where do you see date field?

Batch UUID is part of `operation`.

laxmikant99 replied:
i saw in mentioned log configuration . `<pattern>%-5level [%thread] %date{ISO8601} %F:%L - %msg%n</pattern>`.. but i think that is just an example and can be configured by user accordingly..

vinaykumarchella replied:
Correct.

laxmikant99 commented on src/java/org/apache/cassandra/audit/AuditLogEntryType.java:
```diff
@@ -0,0 +1,87 @@
+    LIST_PERMISSIONS("DCL"),
+    ALTER_TYPE("DDL"),
+    CREATE_ROLE("DCL"),
+    USE_KS("DDL"),
```
USE_KS should be kept in OTHER category rather than DDL?

vinaykumarchella replied:
Sure

laxmikant99 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -0,0 +1,323 @@
+    {
+        if (isAuditingEnabled()
+            && (null != logEntry)
+            && ((null == logEntry.getKeyspace()) || !isSystemKeyspace(logEntry.getKeyspace()))
```
shouldn't we audit system keyspaces as well ?

vinaykumarchella replied:
System keyspaces are mostly accessed through C* internally.

laxmikant99 replied:
is there a way to figure out that query on system keyspces are internal or external?

vinaykumarchella replied:
I am not aware of that.

laxmikant99 commented on src/java/org/apache/cassandra/audit/AuditLogEntryType.java:
```diff
@@ -0,0 +1,87 @@
+    LOGIN_SUCCESS("AUTH");
+
+
+    final String category;
```
shouldn't category be also enum ?

vinaykumarchella replied:
Category is not being used anywhere else apart from within `AuditLogEntry`, Good to avoid unnecessarily types.

vinaykumarchella committed:
Stefan review comments: Fixed isFiltered logic, added more test cases, remove unused error() from IAuditLogger

vinaykumarchella committed:
Fixed USE_KS AuditLogEntryType

vinaykumarchella committed:
circleci high_capacity_env

jasobrown committed:
- cleaned up whitespace and some comments
- made AuditLogFilters atomic and final, and no longer a singleton. a volatile reference is maintained in AuditLogManager
- moved getLogEntry methods from AuditLogManager to AuditLogEntry

vinaykumarchella committed:
Fixes to failing tests in AuditLogFilter

vinaykumarchella committed:
Implement nodetool commands to enable/disable auditlog

vinaykumarchella committed:
Cleanup white spaces

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -45,30 +46,29 @@
-            this.auditLogger = getAuditLogger();
+            logger.info("Audit logging is enabled.");
+            this.auditLogger = getAuditLogger(DatabaseDescriptor.getAuditLoggingOptions().logger);
+//            this.isAuditLogEnabled = true;
```
nit: remove this

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -197,8 +198,38 @@ public void logBatch(String batchTypeName, List<Object> queryOrIdList, List<List
+
+        this.auditLogger = getAuditLogger(auditLogOptions.logger);
+
+            /* Ensure oldLogger's stop() is called after we swap it with new logger otherwise,
```
nit: whitespace

vinaykumarchella replied:
👍

vinaykumarchella replied:
👍

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -197,8 +198,38 @@ public void logBatch(String batchTypeName, List<Object> queryOrIdList, List<List
+        if (this.auditLogger != null)
+        {
+            this.auditLogger.stop();
+            this.auditLogger = null;
```
I think this can race with ``log`` (which is not synchronized afaict) and cause a NPE. I think it's reasonable though to just catch NPE in ``log`` since that can only happen if the user has requested logging to stop.

vinaykumarchella replied:
As per our offline discussion, instead of NPE handling in `log`, implemented `NoOpAuditLogger`.

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -197,8 +198,38 @@ public void logBatch(String batchTypeName, List<Object> queryOrIdList, List<List
+
+        if (oldLogger != null && oldLogger.getClass().getSimpleName().equals(auditLogOptions.logger))
+        {
+            logger.info("New AuditLogger [{}] is same as existing logger, hence not initializing the logger", auditLogOptions.logger);
```
Might want to make it clear that filters were still updated. Otherwise as a user they might think "oh it didn't do anything" when it did in fact reload the filters.

vinaykumarchella replied:
👍

vinaykumarchella replied:
StorageService:enableAuditLog() method logs the changes in audit log filters anyways. That would make it clear that even thought logger is not reloaded, filters are.

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -197,8 +198,38 @@ public void logBatch(String batchTypeName, List<Object> queryOrIdList, List<List
+    public synchronized void enableAuditLog(AuditLogOptions auditLogOptions) throws ConfigurationException
     {
-        filter = AuditLogFilter.create();
+        logger.trace("Audit logging is being enabled. Reloading AuditLogOptions.");
```
nit: probably should be at info.

vinaykumarchella replied:
Might be too much at info, will keep at debug. We already have similar statement in StorageService

jolynch commented on src/java/org/apache/cassandra/db/fullquerylog/FullQueryLogger.java:
```diff
@@ -112,18 +126,23 @@ public synchronized void configure(Path path, String rollCycle, boolean blocking
         }
 
-        if (path.toFile().exists())
+        if(cleanDirectory)
```
nit: whitespace

vinaykumarchella replied:
👍

jolynch commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -197,8 +198,38 @@ public void logBatch(String batchTypeName, List<Object> queryOrIdList, List<List
+
+        if (oldLogger != null)
+        {
+            oldLogger.stop();
```
I'm somewhat concerned about the race conditions inherent in this for the ``BinAuditLogger``, for the ``FileAuditLogger`` it seems clear this would be ok. In particular I wonder if ``log`` could raise all sorts of ``RuntimeExceptions`` if ``oldLogger.log`` get's called in ``log`` after this thread has called ``stop``. This may be fine, but just curious if we can at least add a comment here explaining the order of these operations and how that might impact live queries.

vinaykumarchella replied:
As we discussed offline, BinAuditLogger also handles such cases.

jolynch commented on src/java/org/apache/cassandra/utils/FBUtilities.java:
```diff
@@ -520,6 +520,22 @@ public static IAuditLogger newAuditLogger(String className) throws Configuration
         return FBUtilities.construct(className, "Audit logger");
     }
 
+    public static boolean isAuditLoggerClassExists(String className)
```
maybe can you re-use ``FBUtilities.construct`` rather than this method?

vinaykumarchella replied:
`FBUtilities.construct` instantiates an object of a class, hence added this new method

jolynch reviewed (commented):
Overall latest changes look good. Might want to do a search for "if(" though and add in the whitespace per Jason's earlier CR.

vinaykumarchella committed:
Fixed Joe's review comments

vinaykumarchella committed:
Updated documentation for nodetool auditlog commands

jasobrown committed:
first commit of merging FQL and audit logging, and refactoring ALEntry to use the builder pattern

vinaykumarchella committed:
Minor fixes to pass tests and logs from previous commit

vinaykumarchella committed:
Moving category in AuditLogEntryType to AuditLogEntryCategory Enum

vinaykumarchella committed:
Filter unintended messages to FQL, fix duplicate messsage logging to FQL with Batches

vinaykumarchella committed:
Updated AuditLogging documentation for reloadFilters

vinaykumarchella commented on src/java/org/apache/cassandra/audit/AuditLogEntry.java:
```diff
@@ -0,0 +1,302 @@
+            /*
+             * FQLTime is time in milliseconds since the epoch since the operation was invoked, this does not involve operation execution time
+             */
+            long fqlTime = System.currentTimeMillis() - TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - queryStartNanoTime);
```
@jasobrown What do you think on this calculation? Is alternative to directly rely on `queryStartNanoTime` sgty?

vinaykumarchella replied:
@jasobrown ^

jasobrown replied:
ughh, two system calls in one line. let me see what I can do (will be part of a patch of changes I'm pushing up in a short while).

vinaykumarchella committed:
Fix DatabaseDescriptorRefTest for BinLogAuditLogger

jasobrown committed:
fixed a few formatting issues and some comments; minor clarifications in AuditLogManager; cleaned up/reduced a few log statements

jasobrown committed:
trivial clean ups in test dir

vinaykumarchella committed:
Update documentation for AuditLog Appender, added test case for isFiltered

jasobrown committed:
- cleaned up yaml comments, and point users to the docs; else, it's gonna real messy real quick
- trivial clean ups: NEWS.txt, formatting, white space, comments, removed unused methods, fixed test method names in {{AuditLoggerTest}}
- reworked {{AuditLogFilter#create()}} to eliminate garbage collection of all the {{HashSet}}s, which then we pass to {{ImmutableSet#of}}
- Added path checking in {{AuditLogManager}} to ensure when enabling either FQL or audit logging, the path doesn't conflict with the other. This was easy enough to do for the {{BinLog}}-related {{IAuditLogger}}, but I have no idea how to get the directory or file from logback (when using {{FileAuditLogger}}). Since that's a 'spare' implementation, I've punted for now.
- clarified {{AuditLogEntry#timestamp}} to allow {{Builder}} users to explicitly set the value. This allows the FQL paths to be more in tune with their original code (where it would get {{System#currentTimeMillis()}} instead of using {{queryStartNanoTime}}).

vinaykumarchella committed:
Updated documentation for BinAuditLogger configurations, audit_logs_dir configurations

bvinayakumar commented on doc/source/operating/audit_logging.rst:
```diff
@@ -0,0 +1,203 @@
+ - ``host``: Host IP, where the command is being executed
+ - ``source ip address``: Source IP address from where the request initiated
+ - ``source port``: Source port number from where the request initiated
+ - ``timestamp``: unix time stamp
```
Does each node have a separate audit log file? If so, in a multi-node cassandra setup with replication configured will the timestamp be same on each replicated node for the same request?

vinaykumarchella replied:
Yes, audit log file and location are node dependent. Correct, timestamp is when the query is being executed on a given node.
